### PR TITLE
implement index parallel

### PR DESCRIPTION
**Pull request by FrancoisZhang**

### CODE REVIEW

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java [hunk 1]
```diff
@@ -83,6 +83,15 @@ public class FlatFileNodeStoreBuilder {
      */
     static final String PROP_MERGE_TASK_BATCH_SIZE = "oak.indexer.mergeTaskBatchSize";
 
+    /**
```
Comment (FrancoisZhang) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java:94]: @Ewocker those constants maybe better to be moved to FlatFileStoreSplitter instead

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java [hunk 2]
```diff
@@ -0,0 +1,291 @@
...
+        this.infoProvider = new NodeStateNodeTypeInfoProvider(store.getRoot());
+        this.entryReader = new NodeStateEntryReader(indexHelper.getGCBlobStore());
+    }
+
+    public FlatFileStoreSplitter(FlatFileStore ffs, IndexHelper indexHelper, IndexerSupport indexerSupport, long minimumSplitThreshold) {
+        this(ffs, indexHelper, indexerSupport);
+        this.minimumSplitThreshold = minimumSplitThreshold;
+    }
+
+    public Set<IndexDefinition> getIndexDefinitions() throws IOException, CommitFailedException {
```
Comment (FrancoisZhang) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java]: @Ewocker except the split, for others methods not be used outside, it'd better making private in the beginning

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+            long readPos = 0;
+            int outFileIndex = 1;
+            File currentFile = new File(workDir, "split-" + outFileIndex + "-" + getSortedStoreFileName(useZip));
+            BufferedWriter writer = createWriter(currentFile, useZip);
+            splitFlatFiles.add(currentFile);
+
+            String line;
+            int lineCount = 0;
+            while ((line = reader.readLine()) != null) {
+                NodeStateEntry nse = entryReader.read(line);
```
Comment (FrancoisZhang) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java]: @Ewocker this line seems to be not needed

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+        if (splitThreshold < minimumSplitThreshold || splitSize <= 1) {
+            log.info("split is not necessary, skip splitting");
+            splitFlatFileStores.add(ffs);
+            return splitFlatFileStores;
+        }
+
+        Set<IndexDefinition> indexDefinitions = getIndexDefinitions();
+        Set<NodeTypeInfo> splitNodeTypes = getSplitNodeType(indexDefinitions);
+        Set<String>splitNodeTypesName = splitNodeTypes.stream().map(NodeTypeInfo::getNodeTypeName).collect(Collectors.toSet());
+        log.info("split allowed types: {}", splitNodeTypesName);
```
Comment (FrancoisZhang) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java]: @Ewocker a small change is preferred, can we merge above 3 lines code to the getSplitNodeTypeNames?

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java (hunk 1 above)
Comment (Ewocker) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java:94]: This is specifically for merger. Do you mean to move to merger instead?

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java (hunk 3 above)
Comment (Ewocker) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java]: This was used for keeping track of parent stack, was gonna remove, but now we actually might need it.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java (hunk 4 above)
Comment (Ewocker) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java]: updated

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java (hunk 2 above)
Comment (Ewocker) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreSplitter.java]: updated

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java [hunk 5]
```diff
@@ -42,13 +48,15 @@ public class FlatFileNodeStoreBuilder {
 
     private static final String FLAT_FILE_STORE_DIR_NAME_PREFIX = "flat-fs-";
 
-    public static final String OAK_INDEXER_USE_ZIP = "oak.indexer.useZip";
+    public static final String OAK_INDEXER_USE_ZIP = "oak.indexer.compressionEnabled";
```
Comment (FrancoisZhang) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java]: @Ewocker might we should not change the property key here for risk of compatibility if which is used somewhere already.

Comment (Ewocker) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileNodeStoreBuilder.java]: done

**Pull request merged**